source/windows/bcrypt_rsa.c

```diff
@@ -0,0 +1,458 @@
+
+    if (!BCRYPT_SUCCESS(status)) {
+        return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+    }
+
+    out->len += length_written;
+    return AWS_OP_SUCCESS;
+}
+
+int s_rsa_decrypt(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm,
+    struct aws_byte_cursor ciphertext,
+    struct aws_byte_buf *out) {
+    struct bcrypt_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    if (s_is_not_supported_enc_algo(algorithm)) {
+        return AWS_OP_ERR;
+    }
+
+    if (ciphertext.len == 0) {
+        return AWS_OP_SUCCESS;
+    }
+
+    BCRYPT_OAEP_PADDING_INFO padding_info_oaep;
+    padding_info_oaep.pszAlgId =
+        algorithm == AWS_CAL_RSA_ENCRYPTION_OAEP_SHA256 ? BCRYPT_SHA256_ALGORITHM : BCRYPT_SHA512_ALGORITHM;
+    padding_info_oaep.pbLabel = NULL;
+    padding_info_oaep.cbLabel = 0;
+
+    ULONG length_written = 0;
+    NTSTATUS status = BCryptDecrypt(
+        key_pair_impl->key_handle,
+        ciphertext.ptr,
+        (ULONG)ciphertext.len,
+        algorithm == AWS_CAL_RSA_ENCRYPTION_PKCS1_5 ? NULL : &padding_info_oaep,
+        NULL,
+        0,
+        out->buffer + out->len,
+        (ULONG)(out->capacity - out->len),
+        &length_written,
+        algorithm == AWS_CAL_RSA_ENCRYPTION_PKCS1_5 ? BCRYPT_PAD_PKCS1 : BCRYPT_PAD_OAEP);
+
+    if (!BCRYPT_SUCCESS(status)) {
+        return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+    }
+
+    out->len += length_written;
+    return AWS_OP_SUCCESS;
+}
+
+/*
+ * Allocates and fills out appropriate padding info for algo. Up to caller to destroy.
+ */
+void *s_create_sign_padding_info(struct aws_allocator *allocator, enum aws_rsa_signature_algorithm algorithm) {
+    if (algorithm == AWS_CAL_RSA_SIGNATURE_PKCS1_5_SHA256) {
+        BCRYPT_PKCS1_PADDING_INFO *padding_info = aws_mem_calloc(allocator, 1, sizeof(BCRYPT_PKCS1_PADDING_INFO));
+        padding_info->pszAlgId = BCRYPT_SHA256_ALGORITHM;
+        return padding_info;
+    } else if (algorithm == AWS_CAL_RSA_SIGNATURE_PSS_SHA256) {
+        BCRYPT_PSS_PADDING_INFO *padding_info = aws_mem_calloc(allocator, 1, sizeof(BCRYPT_PSS_PADDING_INFO));
+        padding_info->pszAlgId = BCRYPT_SHA256_ALGORITHM;
+        padding_info->cbSalt = 32;
+        return padding_info;
+    }
+
+    return NULL;
+}
+
+int s_rsa_sign(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signature_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_buf *out) {
+    struct bcrypt_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    void *padding_info = s_create_sign_padding_info(key_pair->allocator, algorithm);
+    if (padding_info == NULL) {
+        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    }
+
+    size_t output_buf_space = out->capacity - out->len;
+
+    if (output_buf_space < aws_rsa_key_pair_signature_length(key_pair)) {
```

Comment:
should we just have the parent aws_rsa_key_pair_sign_message() implementation do this common check?